File: examples/security/encryption/DecryptTest.java


import java.io.*;
import javax.crypto.Cipher;
import javax.crypto.KeyGenerator;
import javax.crypto.NoSuchPaddingException;
import javax.crypto.IllegalBlockSizeException;
import javax.crypto.BadPaddingException;
import java.security.Key;
import java.security.Security;
import java.security.NoSuchAlgorithmException;
import java.security.InvalidKeyException;

public class DecryptTest
{

    public static void main(String[] args) 
	{

        try {
			File dataFile = new File("data.encrypted");
			FileInputStream data = new FileInputStream(dataFile);
			FileInputStream fis = new FileInputStream("key");
			ObjectInputStream ois = new ObjectInputStream(fis);

            Key key = (Key) ois.readObject();
			ois.close();

            /*Cipher cipher = Cipher.getInstance("DES");*/
            Cipher cipher = Cipher.getInstance("AES");

			byte [] result = new byte[(int)dataFile.length()];
			int n = data.read(result);
			System.out.println("read "+n+" bytes");
			data.close();

            cipher.init(Cipher.DECRYPT_MODE, key);
            byte[] original = cipher.doFinal(result);
            System.out.println("Decrypted data: " + new String(original));

        } catch (ClassNotFoundException e) {
            e.printStackTrace();
        } catch (IOException e) {
            e.printStackTrace();
        } catch (NoSuchAlgorithmException e) {
            e.printStackTrace();
        } catch (NoSuchPaddingException e) {
            e.printStackTrace();
        } catch (InvalidKeyException e) {
            e.printStackTrace();
        } catch (IllegalStateException e) {
            e.printStackTrace();
        } catch (IllegalBlockSizeException e) {
            e.printStackTrace();
        } catch (BadPaddingException e) {
            e.printStackTrace();
        }

    }

}

File: examples/golang/security/encryption/DecryptTest.go
package main

import (
	"crypto/aes"
	"crypto/cipher"
	"fmt"
	"os"
)

func check(e error) {
	if e != nil {
		panic(e)
	}
}

func main() {
	//Open the encrypted file and defer the closing of it until the end of the function
	dataFile, err := os.Open("./data.encrypted")
	check(err)
	defer dataFile.Close()

	//Open the key file and defer the closing of it until the end of the function
	fis, err := os.Open("./key")
	check(err)
	defer fis.Close()

	//Read the key from the file
	keyBytes := make([]byte, 32)
	_, err = fis.Read(keyBytes)
	fmt.Println("Key: ", string(keyBytes))

	//Read the encrypted data from the file
	enc := make([]byte, 4096)
	numRead, err := dataFile.Read(enc)
	fmt.Println("Read ", numRead, " bytes")
	check(err)
	enc = enc[:numRead]

	//Create the cipher block from the key
	block, err := aes.NewCipher(keyBytes)
	check(err)

	//Create the GCM cipher from the cipher block
	aesGCM, err := cipher.NewGCM(block)
	check(err)

	//Get the nonce size from the GCM cipher
	nonceSize := aesGCM.NonceSize()
	//Get the nonce from the encrypted data and the ciphertext
	nonce, ciphertext := enc[:nonceSize], enc[nonceSize:]

	//Decrypt the data
	plaintext, err := aesGCM.Open(nil, nonce, ciphertext, nil)
	check(err)
	fmt.Print("Decrypted data: ")
	fmt.Println(string(plaintext))
}
